File: Swift/microgpt.swift
import Foundation

func random_seed(_ seed: Int) {
    srand48(seed)
}

func random_gauss(mean: Double = 0.0, std: Double = 1.0) -> Double {
    // Box-Muller transform
    let u1 = max(1e-12, drand48())
    let u2 = drand48()
    let z0 = sqrt(-2.0 * log(u1)) * cos(2.0 * Double.pi * u2)
    return mean + std * z0
}

func random_shuffle<T>(_ array: inout [T]) {
    guard array.count > 1 else { return }
    for i in stride(from: array.count - 1, through: 1, by: -1) {
        let j = Int(drand48() * Double(i + 1))
        array.swapAt(i, j)
    }
}

func random_choices(weights: [Double]) -> Int {
    let total = weights.reduce(0.0, +)
    let r = drand48() * total
    var cumsum = 0.0
    for (i, w) in weights.enumerated() {
        cumsum += w
        if r <= cumsum {
            return i
        }
    }
    return weights.count - 1
}

random_seed(42)

let input_path = "input.txt"
var docs = try! String(contentsOfFile: input_path, encoding: .utf8)
    .split(separator: "\n")
    .map { String($0).trimmingCharacters(in: .whitespaces) }
    .filter { !$0.isEmpty }
random_shuffle(&docs)
print("num docs: \(docs.count)")

let uchars = Array(Set(docs.joined())).sorted()
let char_to_token = Dictionary(uniqueKeysWithValues: uchars.enumerated().map { ($0.element, $0.offset) })
let BOS = uchars.count
let vocab_size = uchars.count + 1
print("vocab size: \(vocab_size)")

class Value {
    var data: Double
    var grad: Double
    var _children: [Value]
    var _local_grads: [Double]

    init(_ data: Double, children: [Value] = [], local_grads: [Double] = []) {
        self.data = data
        self.grad = 0.0
        self._children = children
        self._local_grads = local_grads
    }

    static func + (lhs: Value, rhs: Value) -> Value {
        return Value(lhs.data + rhs.data, children: [lhs, rhs], local_grads: [1.0, 1.0])
    }

    static func + (lhs: Value, rhs: Double) -> Value {
        return lhs + Value(rhs)
    }

    static func + (lhs: Double, rhs: Value) -> Value {
        return Value(lhs) + rhs
    }

    static func * (lhs: Value, rhs: Value) -> Value {
        return Value(lhs.data * rhs.data, children: [lhs, rhs], local_grads: [rhs.data, lhs.data])
    }

    static func * (lhs: Value, rhs: Double) -> Value {
        return lhs * Value(rhs)
    }

    static func * (lhs: Double, rhs: Value) -> Value {
        return Value(lhs) * rhs
    }

    func pow(_ other: Double) -> Value {
        return Value(
            Foundation.pow(data, other), children: [self],
            local_grads: [other * Foundation.pow(data, other - 1.0)])
    }

    func log() -> Value {
        return Value(Foundation.log(data), children: [self], local_grads: [1.0 / data])
    }

    func exp() -> Value {
        return Value(Foundation.exp(data), children: [self], local_grads: [Foundation.exp(data)])
    }

    func relu() -> Value {
        return Value(max(0.0, data), children: [self], local_grads: [data > 0 ? 1.0 : 0.0])
    }

    static prefix func - (x: Value) -> Value {
        return x * -1.0
    }

    static func - (lhs: Value, rhs: Value) -> Value {
        return lhs + (-rhs)
    }

    static func - (lhs: Double, rhs: Value) -> Value {
        return Value(lhs) + (-rhs)
    }

    static func - (lhs: Value, rhs: Double) -> Value {
        return lhs + (-Value(rhs))
    }

    static func / (lhs: Value, rhs: Value) -> Value {
        return lhs * rhs.pow(-1.0)
    }

    static func / (lhs: Value, rhs: Double) -> Value {
        return lhs * Value(rhs).pow(-1.0)
    }

    static func / (lhs: Double, rhs: Value) -> Value {
        return Value(lhs) * rhs.pow(-1.0)
    }

    func backward() {
        var topo: [Value] = []
        var visited = Set<ObjectIdentifier>()
        func build_topo(_ v: Value) {
            let id = ObjectIdentifier(v)
            if !visited.contains(id) {
                visited.insert(id)
                for child in v._children {
                    build_topo(child)
                }
                topo.append(v)
            }
        }
        build_topo(self)
        self.grad = 1.0
        for v in topo.reversed() {
            for (child, local_grad) in zip(v._children, v._local_grads) {
                child.grad += local_grad * v.grad
            }
        }
    }
}

let n_embd = 16
let n_head = 4
let n_layer = 1
let block_size = 16
let head_dim = n_embd / n_head

func matrix(_ nout: Int, _ nin: Int, std: Double = 0.08) -> [[Value]] {
    return (0..<nout).map { _ in
        (0..<nin).map { _ in Value(random_gauss(mean: 0.0, std: std)) }
    }
}

var state_dict: [String: [[Value]]] = [
    "wte": matrix(vocab_size, n_embd),
    "wpe": matrix(block_size, n_embd),
    "lm_head": matrix(vocab_size, n_embd),
]
for i in 0..<n_layer {
    state_dict["layer\(i).attn_wq"] = matrix(n_embd, n_embd)
    state_dict["layer\(i).attn_wk"] = matrix(n_embd, n_embd)
    state_dict["layer\(i).attn_wv"] = matrix(n_embd, n_embd)
    state_dict["layer\(i).attn_wo"] = matrix(n_embd, n_embd)
    state_dict["layer\(i).mlp_fc1"] = matrix(4 * n_embd, n_embd)
    state_dict["layer\(i).mlp_fc2"] = matrix(n_embd, 4 * n_embd)
}
var params: [Value] = []
func append_params(_ m: [[Value]]) {
    for row in m {
        params.append(contentsOf: row)
    }
}
append_params(state_dict["wte"]!)
append_params(state_dict["wpe"]!)
append_params(state_dict["lm_head"]!)
for i in 0..<n_layer {
    append_params(state_dict["layer\(i).attn_wq"]!)
    append_params(state_dict["layer\(i).attn_wk"]!)
    append_params(state_dict["layer\(i).attn_wv"]!)
    append_params(state_dict["layer\(i).attn_wo"]!)
    append_params(state_dict["layer\(i).mlp_fc1"]!)
    append_params(state_dict["layer\(i).mlp_fc2"]!)
}
print("num params: \(params.count)")

func linear(_ x: [Value], _ w: [[Value]]) -> [Value] {
    return w.map { wo in
        zip(wo, x).map(*).reduce(Value(0.0), +)
    }
}

func softmax(_ logits: [Value]) -> [Value] {
    let max_val = logits.map { $0.data }.max()!
    let exps = logits.map { ($0 - max_val).exp() }
    let total = exps.reduce(Value(0.0), +)
    return exps.map { $0 / total }
}

func rmsnorm(_ x: [Value]) -> [Value] {
    let ms = x.map { $0 * $0 }.reduce(Value(0.0), +) / Double(x.count)
    let scale = (ms + 1e-5).pow(-0.5)
    return x.map { $0 * scale }
}

func gpt(_ token_id: Int, _ pos_id: Int, _ keys: inout [[[Value]]], _ values: inout [[[Value]]])
    -> [Value]
{
    let tok_emb = state_dict["wte"]![token_id]
    let pos_emb = state_dict["wpe"]![pos_id]
    var x = zip(tok_emb, pos_emb).map(+)
    x = rmsnorm(x)

    for li in 0..<n_layer {
        // 1) Multi-head attention block
        let x_residual = x
        x = rmsnorm(x)
        let q = linear(x, state_dict["layer\(li).attn_wq"]!)
        let k = linear(x, state_dict["layer\(li).attn_wk"]!)
        let v = linear(x, state_dict["layer\(li).attn_wv"]!)
        keys[li].append(k)
        values[li].append(v)
        var x_attn: [Value] = []
        for h in 0..<n_head {
            let hs = h * head_dim
            var attn_logits: [Value] = []
            for t in 0..<keys[li].count {
                var dot = Value(0.0)
                for j in 0..<head_dim {
                    dot = dot + q[hs + j] * keys[li][t][hs + j]
                }
                attn_logits.append(dot / Foundation.pow(Double(head_dim), 0.5))
            }
            let attn_weights = softmax(attn_logits)
            let head_out = (0..<head_dim).map { j -> Value in
                var out = Value(0.0)
                for t in 0..<values[li].count {
                    out = out + attn_weights[t] * values[li][t][hs + j]
                }
                return out
            }
            x_attn.append(contentsOf: head_out)
        }
        x = linear(x_attn, state_dict["layer\(li).attn_wo"]!)
        x = zip(x, x_residual).map(+)
        // 2) MLP block
        let x_residual2 = x
        x = rmsnorm(x)
        x = linear(x, state_dict["layer\(li).mlp_fc1"]!)
        x = x.map { $0.relu() }
        x = linear(x, state_dict["layer\(li).mlp_fc2"]!)
        x = zip(x, x_residual2).map(+)
    }

    let logits = linear(x, state_dict["lm_head"]!)
    return logits
}

let learning_rate = 0.01
let beta1 = 0.85
let beta2 = 0.99
let eps_adam = 1e-8
var m = Array(repeating: 0.0, count: params.count)
var v = Array(repeating: 0.0, count: params.count)

let num_steps = 1000
for step in 0..<num_steps {

    let doc = docs[step % docs.count]
    let tokens = [BOS] + doc.map { char_to_token[$0]! } + [BOS]
    let n = min(block_size, tokens.count - 1)

    var keys = Array(repeating: [[Value]](), count: n_layer)
    var values = Array(repeating: [[Value]](), count: n_layer)
    var losses: [Value] = []
    for pos_id in 0..<n {
        let token_id = tokens[pos_id]
        let target_id = tokens[pos_id + 1]
        let logits = gpt(token_id, pos_id, &keys, &values)
        let probs = softmax(logits)
        let loss_t = -probs[target_id].log()
        losses.append(loss_t)
    }
    let loss = (1.0 / Double(n)) * losses.reduce(Value(0.0), +)

    loss.backward()

    let lr_t = learning_rate * (1.0 - Double(step) / Double(num_steps))
    for i in 0..<params.count {
        let p = params[i]
        m[i] = beta1 * m[i] + (1.0 - beta1) * p.grad
        v[i] = beta2 * v[i] + (1.0 - beta2) * p.grad * p.grad
        let m_hat = m[i] / (1.0 - Foundation.pow(beta1, Double(step + 1)))
        let v_hat = v[i] / (1.0 - Foundation.pow(beta2, Double(step + 1)))
        p.data -= lr_t * m_hat / (Foundation.pow(v_hat, 0.5) + eps_adam)
        p.grad = 0.0
    }

    fputs(String(format: "\rstep %4d / %4d | loss %.4f", step + 1, num_steps, loss.data), stdout)
    fflush(stdout)
    if step + 1 == num_steps { fputs("\n", stdout) }
}

let temperature = 0.5
print("\n--- inference (new, hallucinated names) ---")
for sample_idx in 0..<20 {
    var keys = Array(repeating: [[Value]](), count: n_layer)
    var values = Array(repeating: [[Value]](), count: n_layer)
    var token_id = BOS
    var sample: [Character] = []
    for pos_id in 0..<block_size {
        let logits = gpt(token_id, pos_id, &keys, &values)
        let probs = softmax(logits.map { $0 / temperature })
        token_id = random_choices(weights: probs.map { $0.data })
        if token_id == BOS {
            break
        }
        sample.append(uchars[token_id])
    }
    print(String(format: "sample %2d: %@", sample_idx + 1, String(sample)))
}

File: Rust/microgpt.rs
use std::collections::HashMap;
use std::f64::consts::PI;
use std::fs;
use std::io::{self, Write};

const N_EMBD: usize = 16;
const N_HEAD: usize = 4;
const N_LAYER: usize = 1;
const BLOCK_SIZE: usize = 16;
const NUM_STEPS: usize = 1000;
const HEAD_DIM: usize = N_EMBD / N_HEAD;

struct Rng {
    state: u64,
}

impl Rng {
    fn new(seed: u64) -> Self {
        let state = if seed == 0 { 0x9E3779B97F4A7C15 } else { seed };
        Self { state }
    }

    fn next_u64(&mut self) -> u64 {
        let mut x = self.state;
        x ^= x >> 12;
        x ^= x << 25;
        x ^= x >> 27;
        self.state = x;
        x.wrapping_mul(0x2545_F491_4F6C_DD1D)
    }

    fn next_f64(&mut self) -> f64 {
        let r = self.next_u64() >> 11;
        r as f64 * (1.0 / 9007199254740992.0)
    }

    fn gauss(&mut self, mean: f64, std: f64) -> f64 {
        let u1 = (1.0 - self.next_f64()).max(f64::MIN_POSITIVE);
        let u2 = self.next_f64();
        let z = (-2.0 * u1.ln()).sqrt() * (2.0 * PI * u2).cos();
        mean + std * z
    }

    fn shuffle<T>(&mut self, arr: &mut [T]) {
        if arr.len() <= 1 {
            return;
        }
        for i in (1..arr.len()).rev() {
            let j = (self.next_u64() % ((i + 1) as u64)) as usize;
            arr.swap(i, j);
        }
    }

    fn choices(&mut self, weights: &[f64]) -> usize {
        assert!(!weights.is_empty(), "choices() received empty weights");
        let total: f64 = weights.iter().sum();
        let mut r = self.next_f64() * total;
        for (i, w) in weights.iter().enumerate() {
            r -= *w;
            if r <= 0.0 {
                return i;
            }
        }
        weights.len() - 1
    }
}

struct Tape {
    data: Vec<f64>,
    grad: Vec<f64>,
    edge_start: Vec<usize>,
    edge_count: Vec<usize>,
    visit_id: Vec<u32>,
    edge_to: Vec<usize>,
    edge_w: Vec<f64>,
    node_size: usize,
    edge_size: usize,
    visit_counter: u32,
}

impl Tape {
    fn new() -> Self {
        Self {
            data: Vec::new(),
            grad: Vec::new(),
            edge_start: Vec::new(),
            edge_count: Vec::new(),
            visit_id: Vec::new(),
            edge_to: Vec::new(),
            edge_w: Vec::new(),
            node_size: 0,
            edge_size: 0,
            visit_counter: 0,
        }
    }

    fn val(&mut self, data: f64) -> usize {
        let idx = self.node_size;
        self.node_size += 1;
        if idx < self.data.len() {
            self.data[idx] = data;
            self.grad[idx] = 0.0;
            self.edge_start[idx] = self.edge_size;
            self.edge_count[idx] = 0;
            self.visit_id[idx] = 0;
        } else {
            self.data.push(data);
            self.grad.push(0.0);
            self.edge_start.push(self.edge_size);
            self.edge_count.push(0);
            self.visit_id.push(0);
        }
        idx
    }

    fn add_child(&mut self, parent: usize, child: usize, weight: f64) {
        if self.edge_size < self.edge_to.len() {
            self.edge_to[self.edge_size] = child;
            self.edge_w[self.edge_size] = weight;
        } else {
            self.edge_to.push(child);
            self.edge_w.push(weight);
        }
        self.edge_size += 1;
        self.edge_count[parent] += 1;
    }

    fn reset(&mut self, base_nodes: usize) {
        self.node_size = base_nodes;
        self.edge_size = 0;
        for i in 0..base_nodes {
            self.grad[i] = 0.0;
            self.edge_start[i] = 0;
            self.edge_count[i] = 0;
            self.visit_id[i] = 0;
        }
    }
}

fn add(tape: &mut Tape, lhs: usize, rhs: usize) -> usize {
    let out = tape.val(tape.data[lhs] + tape.data[rhs]);
    tape.add_child(out, lhs, 1.0);
    tape.add_child(out, rhs, 1.0);
    out
}

fn mul(tape: &mut Tape, lhs: usize, rhs: usize) -> usize {
    let out = tape.val(tape.data[lhs] * tape.data[rhs]);
    tape.add_child(out, lhs, tape.data[rhs]);
    tape.add_child(out, rhs, tape.data[lhs]);
    out
}

fn pow_op(tape: &mut Tape, base: usize, exponent: f64) -> usize {
    let base_data = tape.data[base];
    let out = tape.val(base_data.powf(exponent));
    tape.add_child(out, base, exponent * base_data.powf(exponent - 1.0));
    out
}

fn relu_op(tape: &mut Tape, v: usize) -> usize {
    let d = tape.data[v];
    let out = tape.val(d.max(0.0));
    tape.add_child(out, v, if d > 0.0 { 1.0 } else { 0.0 });
    out
}

fn log_op(tape: &mut Tape, v: usize) -> usize {
    let out = tape.val(tape.data[v].ln());
    tape.add_child(out, v, 1.0 / tape.data[v]);
    out
}

fn exp_op(tape: &mut Tape, v: usize) -> usize {
    let d = tape.data[v].exp();
    let out = tape.val(d);
    tape.add_child(out, v, d);
    out
}

fn sub(tape: &mut Tape, lhs: usize, rhs: usize) -> usize {
    let out = tape.val(tape.data[lhs] - tape.data[rhs]);
    tape.add_child(out, lhs, 1.0);
    tape.add_child(out, rhs, -1.0);
    out
}

fn div(tape: &mut Tape, lhs: usize, rhs: usize) -> usize {
    let inv = pow_op(tape, rhs, -1.0);
    mul(tape, lhs, inv)
}

fn div_f(tape: &mut Tape, lhs: usize, rhs: f64) -> usize {
    let inv = tape.val(1.0 / rhs);
    mul(tape, lhs, inv)
}

fn build_topo(tape: &mut Tape, v: usize, topo: &mut Vec<usize>, visit_id: u32) {
    if tape.visit_id[v] == visit_id {
        return;
    }
    tape.visit_id[v] = visit_id;
    let start = tape.edge_start[v];
    let count = tape.edge_count[v];
    for i in 0..count {
        build_topo(tape, tape.edge_to[start + i], topo, visit_id);
    }
    topo.push(v);
}

fn backward(tape: &mut Tape, root: usize) {
    tape.visit_counter = tape.visit_counter.wrapping_add(1);
    if tape.visit_counter == 0 {
        tape.visit_counter = 1;
    }
    let visit_id = tape.visit_counter;
    let mut topo = Vec::new();
    build_topo(tape, root, &mut topo, visit_id);

    tape.grad[root] = 1.0;
    for &v in topo.iter().rev() {
        let v_grad = tape.grad[v];
        let start = tape.edge_start[v];
        let count = tape.edge_count[v];
        for i in 0..count {
            let edge_idx = start + i;
            let child = tape.edge_to[edge_idx];
            tape.grad[child] += tape.edge_w[edge_idx] * v_grad;
        }
    }
}

struct Matrix {
    rows: usize,
    cols: usize,
    data: Vec<usize>,
}

impl Matrix {
    fn new(rows: usize, cols: usize) -> Self {
        Self {
            rows,
            cols,
            data: vec![usize::MAX; rows * cols],
        }
    }
}

struct Layer {
    attn_wq: Matrix,
    attn_wk: Matrix,
    attn_wv: Matrix,
    attn_wo: Matrix,
    mlp_fc1: Matrix,
    mlp_fc2: Matrix,
}

struct StateDict {
    wte: Matrix,
    wpe: Matrix,
    lm_head: Matrix,
    layers: Vec<Layer>,
}

fn matrix(rows: usize, cols: usize, rng: &mut Rng, tape: &mut Tape, params: &mut Vec<usize>) -> Matrix {
    let mut m = Matrix::new(rows, cols);
    for i in 0..rows {
        for j in 0..cols {
            let p = tape.val(rng.gauss(0.0, 0.08));
            m.data[i * cols + j] = p;
            params.push(p);
        }
    }
    m
}

fn linear(tape: &mut Tape, x: &[usize], w: &Matrix) -> Vec<usize> {
    assert_eq!(x.len(), w.cols, "linear shape mismatch");
    let mut out = Vec::with_capacity(w.rows);
    for i in 0..w.rows {
        let mut sum_v = tape.val(0.0);
        let row_start = i * w.cols;
        for j in 0..w.cols {
            let prod = mul(tape, w.data[row_start + j], x[j]);
            sum_v = add(tape, sum_v, prod);
        }
        out.push(sum_v);
    }
    out
}

fn softmax(tape: &mut Tape, logits: &[usize]) -> Vec<usize> {
    assert!(!logits.is_empty(), "softmax() received empty logits");
    let mut max_val = f64::NEG_INFINITY;
    for &idx in logits {
        max_val = max_val.max(tape.data[idx]);
    }

    let max_node = tape.val(max_val);
    let mut exps = Vec::with_capacity(logits.len());
    let mut total = tape.val(0.0);
    for &idx in logits {
        let diff = sub(tape, idx, max_node);
        let e = exp_op(tape, diff);
        exps.push(e);
        total = add(tape, total, e);
    }

    exps.into_iter().map(|e| div(tape, e, total)).collect()
}

fn rmsnorm(tape: &mut Tape, x: &[usize]) -> Vec<usize> {
    let mut ss = tape.val(0.0);
    for &xi in x {
        let sq = mul(tape, xi, xi);
        ss = add(tape, ss, sq);
    }
    let inv_n = tape.val(1.0 / x.len() as f64);
    let mean = mul(tape, ss, inv_n);
    let eps = tape.val(1e-5);
    let mean_eps = add(tape, mean, eps);
    let scale_v = pow_op(tape, mean_eps, -0.5);
    x.iter().map(|&xi| mul(tape, xi, scale_v)).collect()
}

type Sequence = Vec<Vec<usize>>;
type KVCache = Vec<Sequence>;

fn gpt(
    tape: &mut Tape,
    token_id: usize,
    pos_id: usize,
    state_dict: &StateDict,
    keys: &mut KVCache,
    values: &mut KVCache,
) -> Vec<usize> {
    let tok_start = token_id * N_EMBD;
    let pos_start = pos_id * N_EMBD;
    let mut x: Vec<usize> = (0..N_EMBD)
        .map(|i| add(tape, state_dict.wte.data[tok_start + i], state_dict.wpe.data[pos_start + i]))
        .collect();
    x = rmsnorm(tape, &x);

    for (li, layer) in state_dict.layers.iter().enumerate() {
        let x_residual = x.clone();
        x = rmsnorm(tape, &x);

        let q = linear(tape, &x, &layer.attn_wq);
        let k = linear(tape, &x, &layer.attn_wk);
        let v = linear(tape, &x, &layer.attn_wv);
        keys[li].push(k.clone());
        values[li].push(v.clone());

        let mut x_attn = Vec::with_capacity(N_EMBD);
        for h in 0..N_HEAD {
            let hs = h * HEAD_DIM;
            let attn_logits: Vec<usize> = keys[li]
                .iter()
                .map(|kt| {
                    let mut dot = tape.val(0.0);
                    for j in 0..HEAD_DIM {
                        let qk = mul(tape, q[hs + j], kt[hs + j]);
                        dot = add(tape, dot, qk);
                    }
                    div_f(tape, dot, (HEAD_DIM as f64).sqrt())
                })
                .collect();

            let attn_weights = softmax(tape, &attn_logits);
            for j in 0..HEAD_DIM {
                let mut out_j = tape.val(0.0);
                for (t, vt) in values[li].iter().enumerate() {
                    let av = mul(tape, attn_weights[t], vt[hs + j]);
                    out_j = add(tape, out_j, av);
                }
                x_attn.push(out_j);
            }
        }

        x = linear(tape, &x_attn, &layer.attn_wo)
            .iter()
            .zip(x_residual.iter())
            .map(|(&a, &b)| add(tape, a, b))
            .collect();

        let x_residual = x.clone();
        x = rmsnorm(tape, &x);
        x = linear(tape, &x, &layer.mlp_fc1);
        for xi in &mut x {
            *xi = relu_op(tape, *xi);
        }
        x = linear(tape, &x, &layer.mlp_fc2)
            .iter()
            .zip(x_residual.iter())
            .map(|(&a, &b)| add(tape, a, b))
            .collect();
    }

    linear(tape, &x, &state_dict.lm_head)
}

fn main() {
    assert_eq!(N_EMBD % N_HEAD, 0, "N_EMBD must be divisible by N_HEAD");
    assert_eq!(HEAD_DIM, N_EMBD / N_HEAD, "HEAD_DIM must match N_EMBD / N_HEAD");

    let mut rng = Rng::new(42);
    let content = fs::read_to_string("input.txt").expect("failed to read input.txt");
    let mut docs: Vec<String> = content
        .lines()
        .map(str::trim)
        .filter(|l| !l.is_empty())
        .map(str::to_string)
        .collect();
    if docs.is_empty() {
        eprintln!("input.txt has no non-empty lines");
        return;
    }
    rng.shuffle(&mut docs);
    println!("num docs: {}", docs.len());

    let mut uchars: Vec<char> = docs.iter().flat_map(|d| d.chars()).collect();
    uchars.sort_unstable();
    uchars.dedup();

    let bos = uchars.len();
    let vocab_size = uchars.len() + 1;
    println!("vocab size: {}", vocab_size);

    let mut stoi = HashMap::new();
    for (i, ch) in uchars.iter().copied().enumerate() {
        stoi.insert(ch, i);
    }

    let mut tape = Tape::new();
    let mut params = Vec::new();

    let mut layers = Vec::new();
    for _ in 0..N_LAYER {
        layers.push(Layer {
            attn_wq: matrix(N_EMBD, N_EMBD, &mut rng, &mut tape, &mut params),
            attn_wk: matrix(N_EMBD, N_EMBD, &mut rng, &mut tape, &mut params),
            attn_wv: matrix(N_EMBD, N_EMBD, &mut rng, &mut tape, &mut params),
            attn_wo: matrix(N_EMBD, N_EMBD, &mut rng, &mut tape, &mut params),
            mlp_fc1: matrix(4 * N_EMBD, N_EMBD, &mut rng, &mut tape, &mut params),
            mlp_fc2: matrix(N_EMBD, 4 * N_EMBD, &mut rng, &mut tape, &mut params),
        });
    }

    let state_dict = StateDict {
        wte: matrix(vocab_size, N_EMBD, &mut rng, &mut tape, &mut params),
        wpe: matrix(BLOCK_SIZE, N_EMBD, &mut rng, &mut tape, &mut params),
        lm_head: matrix(vocab_size, N_EMBD, &mut rng, &mut tape, &mut params),
        layers,
    };

    let num_params = params.len();
    println!("num params: {}", num_params);

    let learning_rate = 0.01;
    let beta1 = 0.85;
    let beta2 = 0.99;
    let eps_adam = 1e-8;
    let mut m = vec![0.0; num_params];
    let mut v = vec![0.0; num_params];

    for step in 0..NUM_STEPS {
        tape.reset(num_params);
        let doc = &docs[step % docs.len()];

        let mut tokens = vec![bos];
        for ch in doc.chars() {
            tokens.push(*stoi.get(&ch).expect("unknown character in doc"));
        }
        tokens.push(bos);

        let n = BLOCK_SIZE.min(tokens.len() - 1);
        let mut keys: KVCache = vec![Vec::new(); N_LAYER];
        let mut values: KVCache = vec![Vec::new(); N_LAYER];

        let mut losses = Vec::with_capacity(n);
        for pos_id in 0..n {
            let token_id = tokens[pos_id];
            let target_id = tokens[pos_id + 1];
            let logits = gpt(
                &mut tape,
                token_id,
                pos_id,
                &state_dict,
                &mut keys,
                &mut values,
            );
            let probs = softmax(&mut tape, &logits);
            let log_p = log_op(&mut tape, probs[target_id]);
            let neg_one = tape.val(-1.0);
            let loss_t = mul(&mut tape, log_p, neg_one);
            losses.push(loss_t);
        }

        let mut loss = tape.val(0.0);
        for &loss_t in &losses {
            loss = add(&mut tape, loss, loss_t);
        }
        let inv_n = tape.val(1.0 / n as f64);
        loss = mul(&mut tape, loss, inv_n);
        let loss_data = tape.data[loss];

        backward(&mut tape, loss);

        let lr_t = learning_rate * (1.0 - step as f64 / NUM_STEPS as f64);
        let t = (step + 1) as i32;
        for (i, &p) in params.iter().enumerate() {
            let g = tape.grad[p];
            m[i] = beta1 * m[i] + (1.0 - beta1) * g;
            v[i] = beta2 * v[i] + (1.0 - beta2) * g * g;
            let m_hat = m[i] / (1.0 - beta1.powi(t));
            let v_hat = v[i] / (1.0 - beta2.powi(t));
            tape.data[p] -= lr_t * m_hat / (v_hat.sqrt() + eps_adam);
            tape.grad[p] = 0.0;
        }

        print!(
            "\rstep {:4} / {:4} | loss {:.4}",
            step + 1,
            NUM_STEPS,
            loss_data
        );
        io::stdout().flush().expect("failed to flush stdout");
        if step + 1 == NUM_STEPS {
            println!();
        }
    }

    println!("\n--- inference (new, hallucinated names) ---");
    let temperature = 0.5;
    for sample_idx in 0..20 {
        tape.reset(num_params);
        let mut keys: KVCache = vec![Vec::new(); N_LAYER];
        let mut values: KVCache = vec![Vec::new(); N_LAYER];
        let mut token_id = bos;
        let mut sample = String::new();

        for pos_id in 0..BLOCK_SIZE {
            let logits = gpt(
                &mut tape,
                token_id,
                pos_id,
                &state_dict,
                &mut keys,
                &mut values,
            );
            let scaled_logits: Vec<usize> = logits
                .iter()
                .map(|&l| div_f(&mut tape, l, temperature))
                .collect();
            let probs = softmax(&mut tape, &scaled_logits);
            let weights: Vec<f64> = probs.iter().map(|&i| tape.data[i]).collect();
            token_id = rng.choices(&weights);

            if token_id == bos {
                break;
            }
            sample.push(uchars[token_id]);
        }

        println!("sample {:2}: {}", sample_idx + 1, sample);
    }
}
